tfjs-backend-webgpu/src/webgpu_program.ts

```diff
@@ -103,23 +103,45 @@ export function getCoordsXYZ(index: number): string {
   }
 }
 
-export function getMainHeaderAndGlobalIndexString(): string {
-  return `
-    ${getMainHeaderString()}
-      let index = getGlobalIndex();
-`;
-}
+export function getMainHeaderString(): string;
+export function getMainHeaderString(index: string): string;
+export function getMainHeaderString(...params: string[]): string {
+  let snippet: string;
+  switch (params.length) {
+    case 0:
+      snippet = `
+        ${getWorkGroupSizeString()}
+        fn _start(@builtin(local_invocation_id) LocalId : vec3<u32>,
+                  @builtin(global_invocation_id) GlobalId : vec3<u32>,
+                  @builtin(num_workgroups) NumWorkgroups : vec3<u32>) {
+          localId = LocalId;
+          globalId = GlobalId;
+          numWorkgroups = NumWorkgroups;
```

Comment:
Can you check if `numWorkgroups = NumWorkgroups;` is still needed for this case? I expect if we use non-linear work group size, we can totally remove the dependency on `num_workgroups` and `getGlobalIndex()`. I am ok to do it in a follow-up PR. Thanks.

Reply:
I don't feel it's safe to assume `getGlobalIndex()` is unused elsewhere. E.g. we currently have this dependency chain in depth-wise conv2d: `main() -> getOutputCoords() -> getGlobalIndex()`.

Reply:
`getGlobalIndex()` is only called when `x.length === outRank` in `getOutputCoords`, which means a flat dispatch layout. Let's keep it for now. Maybe it's not necessary to worry about the very small overhead of  `NumWorkgroups `.

Reply:
I do find the `x.length === outRank` branch is taken while running depthwise_conv2d_test, is this expected?